### ISSUE TITLE
VehicleBlockCollisionEvent returns the wrong block for train carts

### ISSUE THREAD

**Issue by eccentricdevotion** [bergerhealer/TrainCarts#346]
##### Spigot version: git-Paper-168 (MC: 1.15.2)

<hr>

#### Steps to reproduce:
```
List<BlockFace> faces = Arrays.asList(BlockFace.NORTH, BlockFace.WEST, BlockFace.SOUTH, BlockFace.EAST);
    @EventHandler(ignoreCancelled = true)
    public void onVehicleBlockCollision(VehicleBlockCollisionEvent event) {
        if (event.getVehicle() instanceof Minecart) {
            Block block = event.getBlock(); // block in opposite direction of collision
            Bukkit.getLogger().info("event block: " + block.getType().toString());
            Block minecartBlock = event.getVehicle().getLocation().getBlock();
            for (BlockFace face : faces) {
                Block bb = minecartBlock.getRelative(face);
                Bukkit.getLogger().info(face.toString() + ": " + bb.getType().toString());
            }
        }
    }
```
Typically returns something like (where the block the cart collided with is an iron door):
```
[23:26:28 INFO]: event block: RAIL
[23:26:28 INFO]: NORTH: AIR
[23:26:28 INFO]: WEST: IRON_DOOR
[23:26:28 INFO]: SOUTH: AIR
[23:26:28 INFO]: EAST: RAIL
```
I'm guessing this is related to this issue: https://hub.spigotmc.org/jira/browse/SPIGOT-5510, which has been fixed in Spigot, but the entity code for TrainCarts/BKCommonLib has not been updated with the same fix.

**Comment by bbayu123** [comment]
Neither TrainCarts nor BKCommonLib modify the VehicleBlockCollisionEvent, I don't think this is an issue with TrainCarts nor BKCommonLib.

**Comment by eccentricdevotion** [comment]
So why to vanilla minecarts work, and TrainCarts don't?

**Comment by bbayu123** [comment]
Vanilla Minecarts are not related to TrainCarts to begin with, will have to check with dev to see whether that event is being called inside TrainCarts.

**Comment by eccentricdevotion** [comment]
The collision code is not in the event, but in the entity code - net.minecraft.server.Entity.java

**Comment by bbayu123** [comment]
We don't use that entity code.

**Comment by eccentricdevotion** [comment]
Yes, I know, but the TrainCart/BKCommonLib classes extend Bukkit's own Minecart classes, and Minecarts are entities...

**Comment by bbayu123** [comment]
Ok, turns out that we are calculating our own collided block and calling VehicleBlockCollisionEvent. So it must be something wrong with the calculation in BKCommonLib's EntityMoveHandler class

**Comment by bbayu123** [comment]
Related line is this one:
https://github.com/bergerhealer/BKCommonLib/blob/master/src/main/java/com/bergerkiller/bukkit/common/internal/logic/EntityMoveHandler.java#L494

Can you confirm whether it's only 1.15.2 or every single version from 1.8.8 up to 1.15.2 having this problem?

**Comment by eccentricdevotion** [comment]
I only have a 1.15.2 server to test on these days...
The bug only appeared and was fixed in Spigot 1.15, so my guess it's only 1.15.x

**Comment by bbayu123** [comment]
Well, I'm just going to label this and wait for the dev to take over this.

**Comment by bbayu123** [comment]
And no, TrainCarts/BKCommonLib does not extend Bukkit's Minecart classes, the bukkit classes are wrapped into a controller class and bukkit's stuff are thrown out of the window.

**Comment by eccentricdevotion** [comment]
https://github.com/bergerhealer/BKCommonLib/blob/00502668ddb4399a67083e6f4b8fdfd5aadb7b2d/src/main/java/com/bergerkiller/bukkit/common/internal/logic/EntityMoveHandler.java#L482-L490
=>
```
if (d0 > d6) {
   bl_x++;
} else if (d0 < d6) {
   bl_x--;
} else if (d2 > d8) {
   bl_z++;
} else if (d2 < d8) {
   bl_z--;
}
```
Oughta do it...

**Comment by bbayu123** [comment]
The only thing I wonder is whether that change will break many many more stuff than it fixes.

**Comment by bbayu123** [comment]
Also need to make sure it doesn't break 1.8 - 1.14.4

**Comment by eccentricdevotion** [comment]
> Also need to make sure it doesn't break 1.8 - 1.14.4
That'd be what I would worry about, the result of the calculations above are only ever passed to the VehicleBlockCollisionEvent so should be fine for 1.15.x at least.

**Comment by bergerkiller** [comment]
I checked various versions of Minecraft and it looks like this issue is very complicated. I don't know if this is a bug in Bukkit's additions, or mine. My code is how it was up to 1.13.2.

On Minecraft 1.14-1.14.4 it is completely wrong, because EAST and WEST have the exact same conditions:
```
                if (vec3d1.x > vec3d.x) {
                    bl = bl.getRelative(BlockFace.EAST);
                } else if (vec3d.x < vec3d.x) {
                    bl = bl.getRelative(BlockFace.WEST);
```

This same bug perseveres on Minecraft 1.15 and 1.15.1:
```
                if (vec3d1.x > vec3d.x) {
                    bl = bl.getRelative(BlockFace.EAST);
                } else if (vec3d.x < vec3d.x) {
                    bl = bl.getRelative(BlockFace.WEST);
                } else if (vec3d1.z > vec3d.z) {
                    bl = bl.getRelative(BlockFace.SOUTH);
                } else if (vec3d1.z < vec3d.z) {
                    bl = bl.getRelative(BlockFace.NORTH);
                }
```

Then on Minecraft 1.15.2 they "fixed it", but I think they actually introduced a bug that all faces on Vanilla are flipped 180 degrees:
```
                if (vec3d.x > vec3d1.x) {
                    bl = bl.getRelative(BlockFace.EAST);
                } else if (vec3d.x < vec3d1.x) {
                    bl = bl.getRelative(BlockFace.WEST);
                } else if (vec3d.z > vec3d1.z) {
                    bl = bl.getRelative(BlockFace.SOUTH);
                } else if (vec3d.z < vec3d1.z) {
                    bl = bl.getRelative(BlockFace.NORTH);
                }
```
Because if I trace the code, `vec3d` is the d0 in the original code. They've swapped places of the wrong line, they only needed to swap them in the second `else if` but did the other 3 instead.

So basically, in most versions of Minecraft up to today I have done the correct code, because if that d0 and d6 were wrong way around then all the collision physics would be broken.

I don't think vanilla's code is to be trusted right now.

**Comment by bergerkiller** [comment]
So the question is now, do you know it was _wrong_ because it differed from what vanilla minecarts do, or did you know because the block 'collided with' is wrong?

**Comment by bbayu123** [comment]
I mean, if you take a look at the linked Spigot issue that was created.. or was it solved? in Jan 20, spigot claims to have "fixed" it, so.. 🤷🏼‍♂️

**Comment by bergerkiller** [comment]
The comments on that jira ticket are funny to me. "event doesnt fire in this and this situation" etc. Yeah, that's why I rewrote the entire move handler years ago. Kinda late to the party

**Comment by eccentricdevotion** [comment]
My plugin (TARDIS) detects whether a Chest Minecart hits the TARDIS door, if it does and the TARDIS has a "Rail" room the minecart is "teleported" to the TARDIS interior (in a separate world). A user [created an issue](https://github.com/eccentricdevotion/TARDIS/issues/244) asking for support for TrainCarts as in his testing they didn't behave the same way as regular minecarts.

While debugging my code to see why, I logged the block type of the collided block - when it didn't return the expected (Iron Door) - I remembered the Spigot issue and tested the surrounding blocks to see if the same behaviour was apparent.

As far as I know this TARDIS feature has worked correctly with regular minecarts from Minecraft 1.6.2 through to 1.13.2

**Comment by bbayu123** [comment]
So we can be highly certain that 1.14 broke it then? That is count to be very annoying to fix I assume.

**Comment by bergerkiller** [comment]
If all you do in your plugin is check whether the block hit is an iron door, that should be something simple I can check on my end. I'll set something up and check it gives consistent results on all versions of Minecraft.

**Comment by bergerkiller** [comment]
Some fixes, because I did notice that the variable names were incorrect and offset by one. d6 had to be d7 and so on. https://github.com/bergerhealer/BKCommonLib/commit/b2ed5172be6eaba3f97f155b48545183cf3a0032

I did a test setup hitting an iron door and that worked fine on 1.15.2.

**Comment by eccentricdevotion** [comment]
Working for me, thanks.
One question though - what is the best way for me to programmatically spawn a TrainCart cart?

**Comment by bbayu123** [comment]
You could directly take a look a the raw source code of the spawn sign and the tc chest, those are the only 2 ways of spawning traincarts trains

**Comment by bbayu123** [comment]
So, closing this as resolved for the time being as it seems to be fixed... hopefully.
